[영화 극장 선택] 채넛 3, 4단계 미션 제출합니다.

yrsel opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->

### 질문 1. 데이터 베이스 자원 정리는 언제 어떻게 하는 것이 좋을까요 ? 
- 데이터 베이스 같이 생성 비용이 큰 작업은 자원 정리가 중요하다고 생각했는데, 어디서 정리해줘야 할지 고민이 되었고 정리하고 재생성하는 것  또한 비용이 크다고 생각했습니다. 이번 미션에서는 싱글턴으로 단 한번만 생성하도록 강제하고 닫는 것은 앱이 종료될 때 컴파일러에게 맡기게 진행했는데, 더 좋은 방법이 있을까요 ? 혹은 제가 생각한 부분이 잘못된 부분이 있다면 언급해주시면 감사하겠습니다 !

### 질문 2. 프래그먼트 내 리사이클러뷰 어댑터 설정에 대해 궁금한 점이 있습니다.

```kotlin
// BookingHistoryFragment
override fun showHistories(bookingHistories: List<BookedTicketEntity>) {
        activity?.runOnUiThread {
            bookedHistoryAdapter.submitList(bookingHistories)
        }
    }
```
예매 내역에 대해 database로 부터 정보를 가져오는 작업 후 정보를 ui에 넣어주는 작업을 하는 메서드입니다.
기존에는 requireActivity() 를 통해 UiThread를 가져와서 진행했었는데 화면이 전환되는 경우 아직 프래그먼트가 액티비티에 붙지 못한 상황이 발생해 에러를 마주했습니다.
nullable type인 getActivity를 통해 null이 아닐 때에만 화면을 구성해주도록 해 문제를 해결하기는 했으나, 제 생각으론 기존 문제가 프래그먼트가 액티비티에 붙지 못한 상황이었어서 null이 발생해 화면이 전혀 그려지지 않는 경우도 있을 것이라고 생각되었습니다.
다른 더 좋은 방법이 있는 지 궁금합니다, 키워드라도 제시해주시면 감사합니다 :)

### 질문 3. 알림 제거

설정 페이지에서 알림을 설정을 off 시키면 기존에 예약했던 영화 알림을 모두 제거가 필요해보입니다.
검색 해보니 알림을 제거하기 위해서는 PendingIntent 값 전부를 알고 있어야 비교 후 제거 가능하다고 알게 되어 Room 데이터베이스 테이블을 생성해 정보를 저장했어야 됐을 것 같습니다. 
아직 시간이 없어 테이블을 생성하지 못했지만 Room 데이터베이스에 저장 후 설정을 off할 때마다 room의 데이터를 지우고, on할 때마다 예매 내역의 데이터를 받아와 다시 예약을 등록해주는 흐름이 괜찮은걸까요 ?? 
데이터가 커진다면 부하가 큰 작업이 될 것 같아 어떤 방식으로 처리하는 것이 좋은지 궁금합니다.

## 코드 리뷰 커뮤니케이션

안녕하세요, 페로로 🙇‍♂️
푸시 알림 부분을 어떻게 잘 사용해야할까 고민하다보니 제출 시간이 많이 늦어졌습니다
늦었지만 리뷰 잘 부탁드립니다.

- 1, 2단계 리뷰 : Fragment newInstance 메서드의 @JvmStatic 어노테이션 관련 
->  https://github.com/woowacourse/android-movie-theater/pull/112#discussion_r2083832201

## 스크린샷

- 예매 내역

[history.webm](https://github.com/user-attachments/assets/f092656e-2b92-4f97-b977-d99a1452e7f3)

- 퍼미션, 알림

https://github.com/user-attachments/assets/441fa397-7897-4d7b-8aac-513bd4bcee09

yrsel committed:
refactor: BookingActivity xml 스크롤뷰 추가

yrsel commented:
@coderabbitai pause

yrsel committed:
refactor: BookingSeatActivity xml 스크롤뷰 추가

yrsel committed:
refactor: BookingCompleteActivity xml 스크롤뷰 추가

yrsel committed:
docs: README 구현 목록 작성

yrsel committed:
chore: Room 라이브러리 의존성 추가

yrsel committed:
refactor: Seats 요구사항에 불필요한 구현 제거

yrsel committed:
feat: BookedTicketEntity 구현

yrsel committed:
feat: BookedTicket <-> BookedTicketEntity 변환

yrsel committed:
feat: BookedTicketDao 데이터 삽입 기능 추가

yrsel committed:
chore: gson 라이브러리 의존성 추가

- 직렬화, 역직렬화 지원 라이브러리

yrsel committed:
feat: BookedTicketConverter 엔티티 직렬화 추가

yrsel committed:
feat: BookedTicketDatabase 데이터베이스 추가

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
test: BookedTicketDaoTest 데이터 삽입 테스트 추가

yrsel committed:
test: BookedTicketDaoTest 삽입 데이터 일치 여부 테스트 추가

yrsel committed:
test: BookedTicketDaoTest 전체 데이터 조회 테스트 추가

yrsel committed:
feat: 예매 내역 아이템 레이아웃 구현

yrsel committed:
refactor: StringFormatter 시간 변환 로직 수정

yrsel committed:
feat: 예매 내역 아이템 데이터 바인딩 적용

yrsel committed:
feat: BookingHistoryFragment xml 리사이클러뷰 적용

yrsel committed:
feat: HistoryClickListener 예매 내역 리사이클러뷰 리스너 구현

yrsel committed:
feat: BookingHistoryViewHolder 예매 내역 리사이클러뷰 뷰홀더 구현

yrsel committed:
feat: BookingHistoryAdapter 예매 내역 리사이클러뷰 어댑터 구현

yrsel committed:
feat: BookingSeatActivity 예매 완료버튼 클릭 시 데이터베이스 저장 기능 구현

yrsel committed:
feat: BookingHistoryFragment MVP 구조 적용 및 상세 페이지 이동 구현

yrsel committed:
feat: BookingHistoryFragment 리사이클러뷰 구분선 구현

yrsel committed:
chore: MovieBookingActivity 패키지 변경

yrsel committed:
feat: Destination 목적지 상수 구현

yrsel committed:
feat: MovieBookingActivity 팩토리 인텐트 객체 구현

yrsel committed:
refactor: Destination 상수명 수정

yrsel committed:
refactor: 목적지 추가

- BookingHistoryFragment
- BookingSeatActivity

yrsel committed:
refactor: BookingCompleteActivity 목적지 정보와 함께 화면 이동 구현

yrsel committed:
refactor: MovieBookingActivity MVP 구조로 변경

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
feat: 리사이클러뷰 아이템 누르는 효과 추가

yrsel committed:
docs: README 구현 목록 작성

yrsel committed:
refactor: 바텀 네비게이션 레이아웃 수정

yrsel committed:
feat: SettingsFragment xml 푸시알림 뷰 추가

yrsel committed:
refactor: BookingHistoryFragment 화면 전환 시 발생하는 문제 해결

- activity에 fragment가 붙지 못하면서 에러 발생 -> activity가 null이 아닐 때만 실행

yrsel committed:
feat: BookingCompletePresenter 권한 및 알림 설정 흐름 추가

yrsel committed:
feat: BookingCompleteActivity 권한 설정 추가

yrsel committed:
feat: AlarmManagerCompat 버전 호환 도움 클래스 추가

yrsel committed:
feat: BookingCompleteActivity 알람 등록 구현

yrsel committed:
feat: MovieReminderReceiver 영화 알림 수신 역할

yrsel committed:
feat: SettingsFragment 설정 버튼 On/Off 추가 및 상태 유지 구현

yrsel committed:
feat: SettingsFragment 알림 설정 버튼 기기 권한 허용과 동기화 구현

yrsel committed:
feat: SettingsFragment MVP 구조 구현

yrsel committed:
feat: SettingsFragment 설정 허용 변경 시 알림 가능한 영화들 모두 등록

yrsel committed:
feat: BookingCompleteActivity 퍼미션 안받는 버전은 SharedPreferences 값으로 판단

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
docs: README 누락된 체크포인트 업데이트

yrsel committed:
chore: 오타 및 순서 수정

yrsel committed:
refactor: 알람에 대한 권한 추가

yrsel committed:
test: 실패하는 테스트 수정

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
refactor: BookedTicketDao 예매 내역 정렬 추가 구현

yrsel committed:
chore: TheaterBottomSheetDialogFragment 불필요한 어노테이션 제거

laco-dev commented on app/src/main/res/values/preferences.xml:
```diff
@@ -0,0 +1,6 @@
+    <string name="preference_key">THEATER_PREFERENCE_KEY</string>
+    <string name="preference_is_first_notification_request">IS_FIRST_NOTIFICATION_REQUEST</string>
+    <string name="preference_post_notification">POST_NOTIFICATION</string>
+</resources>
```
string resource를 사용하는 이유가 무엇인가요?
아래 내용을 참고해보시고 답변해주시면 좋을 것 같아요
https://woowacourse.slack.com/archives/C08BEPG1SSE/p1744867393313639

yrsel replied:
화면에 보여질 정보를 저장하기 위해서 입니다.
문자열을 상수화할 수 있고 Localize(현지화)하여 다국어를 지원할 수 있습니다.

현재 제가 작성한 값들은 UI에 보여지기 위한 값이 아니기에 적절한 위치가 아닌 것 같습니다 !

laco-dev commented on app/src/main/res/layout/fragment_settings.xml:
```diff
@@ -1,13 +1,54 @@
+            app:layout_constraintTop_toBottomOf="@id/tv_push_notification_title"
+            tools:text="해제하면 푸시 알림을 수신할 수 없습니다." />
+
+        <androidx.appcompat.widget.SwitchCompat
+            android:id="@+id/switch_push_notification"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:layout_marginEnd="16dp"
+            android:checked="@{isEnablePostNotification}"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintTop_toTopOf="parent">
```
SwitchCompat은 Switch와 어떤 차이가 있을까요?

yrsel replied:
SwitchCompat은 Switch와 동일한 기능을 제공하며, thumb 및 track 속성 제어와 같은 추가 기능도 제공합니다.
androidx.appcompat 기반의 switch로 하위 호환성도 제공합니다.
> [SwitchCompat](https://developer.android.com/reference/androidx/appcompat/widget/SwitchCompat.html) provides backward compatibility down to Android 4.0 (API level 14).  (안드로이드 공식문서 Switch 내용 중 일부)

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MovieBookingActivity.kt:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private fun updateBottomNavigation() {
+        val activeFragment =
+            supportFragmentManager.findFragmentById(R.id.main_fragment_container_view)
+                ?: movieListFragment
```
movieListFragment 인스턴스를 미리 변수로 만들어 둔 이유가 있을까요?

yrsel replied:
현재 코드 흐름상 null 발생할 확률이 희박하기에 `movieListFragment`를 지연초기화하여 인스턴스 변수를 선언할 필요가 없었습니다.

`null` 일 때 `MovieListFragment` 인스턴스를 바로 생성하는 방향으로 수정하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MovieBookingActivity.kt:
```diff
@@ -0,0 +1,131 @@
+        }
+    }
+
+    override fun showBottomSheetForHome() {
+        binding.navigation.selectedItemId = R.id.navigation_home
+    }
+
+    override fun showBottomSheetForHistory() {
+        binding.navigation.selectedItemId = R.id.navigation_history
+    }
+
+    override fun showBottomSheetForSettings() {
+        binding.navigation.selectedItemId = R.id.navigation_settings
+    }
```
BottomSheet은 무엇인가요? 어떻게 생긴 녀석인지 궁금합니다.

yrsel replied:
앗, 과거의 저는 어떤 녀석을 원했던 걸 까요...

`MovieBooking`에서 `Presenter`가 필요없어보여 제거하면서 `Navigation`에 대한 내용 수정하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MovieBookingPresenter.kt:
```diff
@@ -0,0 +1,25 @@
+) : MovieBookingContract.Presenter {
+    private lateinit var destination: Destination
+
+    override fun handleDestination(restoreDestination: Destination?) {
```
마지막에 보던 화면이 있었다면 Presenter의 생성자로 받으면 되지 않을까요?

yrsel replied:
`Activity`가 쌓이고 지워지는 구조를 이해하니 `Destination`이 필요없어져 제거하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MovieBookingContract.kt:
```diff
@@ -0,0 +1,23 @@
+        fun handleDestination(restoreDestination: Destination?)
+    }
+
+    interface View {
+        fun showHistory()
+
+        fun showHome()
+
+        fun showSettings()
+
+        fun showBottomSheetForHome()
+
+        fun showBottomSheetForHistory()
+
+        fun showBottomSheetForSettings()
+    }
```
View 내부에서 스스로 호출하는 것들인데 인터페이스로 만들어진 이유가 무엇일까요?
Presenter와 비슷하게 View 내에서도 충분히 최적화가 가능하지 않았을까요?

yrsel replied:
`Presenter`에서 하는 역할이 불필요해 제거하였고 `Activity` 내부에서 `View`에 대한 로직을 직접 처리하도록 하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/data/BookedTicketConverter.kt:
```diff
@@ -0,0 +1,34 @@
+        return LocalDateTime.parse(this, dateTimeFormat)
+    }
+
+    @TypeConverter
+    fun Set<Seat>.toEntityValue(): String {
+        return gson.toJson(this)
+    }
+
+    @TypeConverter
+    fun String.toSet(): Set<Seat> {
+        val type = object : TypeToken<Set<Seat>>() {}.type
+        return gson.fromJson(this, type)
+    }
```
Gson을 이용한 직렬화 좋습니다.

Map이나 Set, List같은 자료구조라면 기본으로 바뀌는 타입? 이 있을것으로 생각했는데요.
아래와 같은 패턴으로 사용해도 동일하게 동작할까요?
```kotlin
Seat.toEntityValue(): String
String.toSeat(): Seat
```

yrsel replied:
Room DB에서는 원시값이 아닌 타입의 경우 `@TypeConverter`와 함께 직렬화, 역직렬화를 해주지 않는다면 알 수 없는 데이터로 인식하기에 작동하지 않을 것 같습니다.

자료구조의 제네릭타입을 사용하지 못하게 된 이유가 엔티티에 도메인 타입을 사용하였기에 발생하는 문제라고 생각합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/data/BookedTicketEntity.kt:
```diff
@@ -0,0 +1,34 @@
+import woowacourse.movie.domain.model.Seats
+import java.time.LocalDateTime
+
+@Entity(tableName = "booked_tickets")
+data class BookedTicketEntity(
+    @PrimaryKey(autoGenerate = true) val id: Long? = null,
+    @ColumnInfo(name = "theater_name") val theaterName: String,
+    @ColumnInfo(name = "movie_title") val movieTitle: String,
+    @ColumnInfo(name = "screening_date_time") val screeningDateTime: LocalDateTime,
+    @ColumnInfo(name = "select_seats") val selectSeats: Set<Seat>,
+    @ColumnInfo(name = "headcount") val headcount: Int,
+)
```
Entity는 domain 모델을 가지고 있어도 되나요?
domain 모델이 데이터베이스를 위한 타입으로 사용되는게 맞을까요?

> 어차피 스펙 똑같은데 그냥 써도 되는거 아닌가?

라고 생각이 드신다면 다른 크루들과도 충분히 논의를 다시 해보세요

yrsel replied:
데이터베이스에 해당 내용을 표현하기 위해서는 `seats`라는 테이블이 별도로 존재하고 `booked_tickets` 테이블과 1(Ticket):N(Seat) 관계로 연결하는 것이 좋을 것 같습니다.

현재 Set\<Seat>가 `gson`과 `@TypeConverter`의 도움을 받아 데이터베이스에 들어갈 때, TEXT 타입 문자열 상태로 들어가게 되어 데이터베이스를 잘 사용하고 있다고 할 수 없을 것 같습니다.
그렇기에 현재 엔티티 같이 도메인 모델을 직접 가지고 있으면 안될 것 같습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/data/BookedTicketDatabase.kt:
```diff
@@ -0,0 +1,35 @@
+            }
+        }
+
+        fun close() {
+            INSTANCE?.close()
+            INSTANCE = null
+        }
```
> 데이터 베이스 같이 생성 비용이 큰 작업은 자원 정리가 중요하다고 생각했는데, 어디서 정리해줘야 할지 고민이 되었고 정리하고 재생성하는 것 또한 비용이 크다고 생각했습니다. 이번 미션에서는 싱글턴으로 단 한번만 생성하도록 강제하고 닫는 것은 앱이 종료될 때 컴파일러에게 맡기게 진행했는데, 더 좋은 방법이 있을까요 ? 혹은 제가 생각한 부분이 잘못된 부분이 있다면 언급해주시면 감사하겠습니다 !

우선 컴파일러에게 맡기다는 표현을 잘못되었습니다.
어플리케이션의 라이프사이클에 맞춰서 자원을 해제하는 것이라면 의미가 있습니다.

실제 프로젝트를 하다 보면 싱글턴으로 관리하는 자원은 매우 많습니다.
지금 깊이 있게 고민할 주제는 아니기에 다른 내용에 집중해보는게 좋을 것 같아요.

laco-dev commented on app/src/androidTest/java/woowacourse/movie/data/BookedTicketDaoTest.kt:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    companion object {
+        private const val ONE_BASED_INDEX_ADJUST_NUMBER = 1
+    }
```
마이너한 내용인데, 이 값을 상수로 처리한 이점은 무엇일까요?

yrsel replied:
테스트 코드에서는 상수화를 하지 않아도 테스트명과 하드코딩된 값을 표현할 변수명으로 충분히 의도를 드러낼 수 있다고 생각합니다.
하지만 이 부분에서는 1-INDEX 부터 시작하는 것을 명시하면 읽는 입장에서 가독성이 조금 좋아지지 않을까하며 작성했습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/notification/MovieReminderReceiver.kt:
```diff
@@ -0,0 +1,84 @@
+        context: Context?,
+        intent: Intent?,
+    ) {
+        context?.let {
```
불필요한 indent를 만들 필요가 없지 않을까요?
early return이나 함수 분리 등 여러 가지 방법이 있을거에요.

그런데 언제 Context가 null이 되는 걸까요?

yrsel replied:
override 함수를 자동완성으로 만들었고 nullable 타입으로 완성시켜주었습니다.
당연히 내부 구현도 nullable 타입일 줄 알았지만 내부 메서드는 자바로 작성되어 있었습니다. 

자바는 nullable, not-null 타입을 구분하지않아 어노테이션으로 not-null 타입을 명시해주지 않았다면 nullable한 타입이 들어올 수 있다고 판단하며 변환해주지 않았을까 하는 추측입니다.

Context가 null이 들어오는 경우가 있는지는 잘 모르겠습니다. 😢

laco-dev commented on app/src/main/java/woowacourse/movie/notification/MovieReminderReceiver.kt:
```diff
@@ -0,0 +1,84 @@
+            NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT)
+        notificationManager.createNotificationChannel(channel)
+
+        val intent = BookingCompleteActivity.newIntent(context, bookedTicket, Destination.HOME)
```
예매 결과 화면으로 반드시 BookedTicket이 있어야만 할까요?
우리는 예매를 완료하면 데이터를 데이터베이스에 저장했을거에요.

그렇다면 예매 결과 화면에서 해당 데이터를 다시 불러오도록 할 수 있지 않을까요?
가령 알림에서 BookedTicket 데이터를 가져오지 못할지도 모릅니다.

yrsel replied:
bookedTicket을 넘기는 방식에서 id값만 넘기고 예매 결과 화면에서 다시 불러오는 방식으로 수정하였습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/notification/MovieReminderReceiver.kt:
```diff
@@ -0,0 +1,84 @@
+        val pendingIntent =
+            PendingIntent.getActivity(
+                context,
+                bookedTicket.hashCode(),
```
hashCode() 값을 넣은 이유는 무엇일까요?

yrsel replied:
bookedTicket 가지고 있는 정보에서 고유하다고 생각되는 값이 없어서 hashCode() 를 구분하기 위한 용도로 사용했었습니다.

이후 bookedTicket에 id 값을 두게되면서 id값을 사용하도록 수정하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompleteActivity.kt:
```diff
@@ -16,16 +28,31 @@ import woowacourse.movie.domain.model.BookedTicket
     BookingCompleteContract.View {
     private val bookingCompletePresenter = BookingCompletePresenter(this)
     private lateinit var binding: ActivityBookingCompleteBinding
+    private val sharedPrefs: SharedPreferences by lazy {
+        this.getSharedPreferences(getString(R.string.preference_key), MODE_PRIVATE)
+    }
+    private val isFirstNotificationRequest: Boolean
+        get() =
+            sharedPrefs.getBoolean(
+                getString(R.string.preference_is_first_notification_request),
+                true,
+            )
+    private val isEnablePostNotification: Boolean
+        get() = sharedPrefs.getBoolean(getString(R.string.preference_post_notification), true)
```
프레젠터는 모델과 협력해서 데이터를 불러오게 됩니다.

그러면 프리퍼런스에 있는 데이터는 무엇일까요?
액티비티가 프리퍼런스와 협력해서 데이터를 불러오는 행위는 괜찮은걸까요?

yrsel replied:
프리퍼런스 값들을 반환하는 별도의 `StorageProvider`라는 객체를 만들어서 값을 전달 할 수 있도록 수정하였습니다

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompleteActivity.kt:
```diff
@@ -71,16 +98,141 @@ class BookingCompleteActivity :
         binding.stringFormatter = StringFormatter
     }
 
+    override fun moveTo(destination: Destination) {
+        val intent =
+            MovieBookingActivity.newIntent(this, destination).apply {
+                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+            }
+        startActivity(intent)
+        finish()
+    }
```
하나의 enum으로 모든 화면 이동을 제어하려고 하지 마세요. 굉장히 안좋은 패턴입니다. 😠

yrsel replied:
넵, 당시에는 문제를 해결하기 위한 하나의 아이디어라고 생각하며 넣었었는데 다시 보니 명확한 의도를 전달하기도 어려울 뿐 더러 페이지가 많아지면 관리할 수 없고 잘못되었음을 느꼈습니다.  Destination 제거하였습니다 ! 😭

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompleteActivity.kt:
```diff
@@ -71,16 +98,141 @@ class BookingCompleteActivity :
+                    Manifest.permission.POST_NOTIFICATIONS,
+                ) == PackageManager.PERMISSION_GRANTED
+
+            if (!isGranted) {
+                if (shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)) {
+                    showRecommendPushNotification()
+                } else { // 최초 요청 || 다시 묻지 않음
+                    if (isFirstNotificationRequest) {
+                        showNotificationDialog(Manifest.permission.POST_NOTIFICATIONS)
+                        sharedPrefs.edit {
+                            putBoolean(
+                                getString(R.string.preference_is_first_notification_request),
+                                false,
+                            )
+                        }
+                    } else {
+                        showRecommendSettingDialog()
+                    }
+                }
+            }
```
객체 생활 체조 원칙을 지켜보죠. 무엇을 고칠 수 있을까요?

yrsel replied:
return을 적절히 사용해 if문의 depth를 줄여보았습니다 !

```
private fun requestPostNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            val hasGranted = hasPermission(Manifest.permission.POST_NOTIFICATIONS)
            if (hasGranted) return

            if (shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)) {
                showPushNotificationRecommend()
                return
            }

            if (StorageProvider.isFirstPostNotificationPermissionRequest) {
                requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                StorageProvider.setFirstPostNotificationPermissionRequestState(false)
            }
        }
    }
```

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompleteActivity.kt:
```diff
@@ -71,16 +98,141 @@ class BookingCompleteActivity :
+            .show()
+    }
+
+    private val requestPermissionLauncher =
+        registerForActivityResult(
+            ActivityResultContracts.RequestPermission(),
+        ) { isChecked ->
+            if (isChecked) {
+                AlarmManagerCompat.requestScheduleExactPermission(this)
+            }
+        }
```
이런 내용은 코드 래빗으로 잡혔어야 하는데, 코드 래빗은 왜 꺼두셨을까요?

yrsel replied:
코드 래빗이 많은 양의 정보를 제공하며 미리 해결방법을 제공하는 것과 리뷰 찾기가 어렵다고 느꼈고 가끔 리뷰어분의 답글에 코드 래빗이 답장을 먼저 달아주는 경우가 있다고 해 코드 래빗을 꺼두었습니다.
페로로의 리뷰에 조금 더 집중하고자 하는 의도였습니다.

yrsel replied:
푸시 알림 권한을 허용했다면 그것에 대한 내용만 처리하고 알람에 대한 권한 설정을 하고 싶은 경우 따로 묻는 방식으로 하는 것이 맞다고 생각되어 수정하였습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompleteActivity.kt:
```diff
@@ -71,16 +98,141 @@ class BookingCompleteActivity :
             else -> super.onOptionsItemSelected(item)
         }
 
+    override fun handlePermission(bookedTicket: BookedTicket) {
+        requestPostNotificationPermission()
+
+        if (hasPostNotification()) {
+            scheduleNotification(bookedTicket)
+        }
+    }
```
예매를 완료한 이후 알림 권한을 강제로 받도록 하는 것이 좋은 흐름일까요?

앱 개발을 하다보면 이러한 UX를 고민하는 연습을 꾸준히 하게 됩니다.
이미 알림 권한을 가지고 있다면 매니저에 등록하는 것은 자연스럽지만 갑자기 알림 권한을 받으라는 팝업을 이 시점에 띄우는 것이 자연스러운지? 고민해보세요!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompletePresenter.kt:
```diff
@@ -17,5 +24,17 @@ class BookingCompletePresenter(
+        if (isAfterBooking()) bookingCompleteView.handlePermission(bookedTicket)
     }
+
+    override fun navigateTo() {
+        when (destination) {
+            Destination.COMPLETE -> bookingCompleteView.moveTo(Destination.HOME)
+            Destination.HISTORY -> bookingCompleteView.moveTo(Destination.HISTORY)
+            else -> bookingCompleteView.moveTo(Destination.HOME)
+        }
+    }
```
홈 화면에서 진입을 했다면 다시 홈 화면을 실행할 필요가 있는 건가요?
이미 뒤에  홈 화면이 깔려있는 상태로 진입하지는 않았을까요?

가령 아래와 같이 진입하더라도 메인 홈 화면이 또 뜨게 됩니다.
> 홈 -> 예매 -> 예매 완료 -> 알림 -> 예매 완료

이 내용은 어려울 수 있지만 실행했던 화면으로 돌아간다는 개념을 현재 방식으로 구현하는 것은 잘못된 방식입니다.

yrsel replied:
이해했습니다 ! , Destination 으로 화면 이동을 제어하려고 한 것의 문제를 잘 보여주는 것 같습니다 😭

laco-dev commented on app/src/main/java/woowacourse/movie/ui/history/contract/BookingHistoryContract.kt:
```diff
@@ -0,0 +1,19 @@
+
+interface BookingHistoryContract {
+    interface Presenter {
+        fun loadBookingHistories(bookedTicketDatabase: BookedTicketDatabase)
```
데이터베이스를 프레젠터가 알고 있어도 되나요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/history/view/BookingHistoryFragment.kt:
```diff
@@ -4,15 +4,68 @@ import android.os.Bundle
+        }
+    }
+
+    override fun showHistories(bookingHistories: List<BookedTicketEntity>) {
+        activity?.runOnUiThread {
+            bookedHistoryAdapter.submitList(bookingHistories)
+        }
+    }
```
> 질문 2. 프래그먼트 내 리사이클러뷰 어댑터 설정에 대해 궁금한 점이 있습니다.

화면이 회전하더라도 showHistories가 불릴 이유가 없는데요.
아직 activity가 초기화가 안된 시점에 어떤 경로로 이 함수가 불리게 되었는지 근본적인 원인은 찾아보셨을까요?

yrsel replied:
당시에는 getActivity(), requireActivity()  생명주기 모두 찍어보았지만 문제 원인을 찾지 못했습니다.

근본적인 원인을 찾기 위해 흐름을 살펴보았습니다.
1. 메인화면에서 프래그먼트들을 replace를 통해 화면을 갈아끼우고 있다.
2. 바텀네비게이션 아이템 선택 리스너에서는  `항상` id값을 비교해서 replace를 진행하고 있었다.
3. 그 중 History Fragment는 `ui thread`에서 데이터를 받아서 그려주고 있다.
4. 화면 전환
5. History Fragment가 있었기에 if분기에서 updateBottomNavigation() 를 수행한다 
6. History 자기 자신으로 replace가 진행될텐데 onViewCreated() 생명주기에서 데이터를 넣어주기 위해 dao에서 데이터를 얻어와서 화면에 보여줘야하는 로직이 thread 내부에서 실행되어 Fragment replace 과정에서 생명주기가 꼬이게 된다. 그래서 가끔 화면 전환시에도 정상 작동하는 경우도 있는 것 같습니다.

해결 방법 : BottomNavigation 아이템 선택 리스너가 호출 될 때 현재 프래그먼트가 자기 자신이라면 return 하도록 변경하였습니다.

--- 
> 화면이 회전하더라도 showHistories가 불릴 이유가 없는데요.

화면이 회전하면 예매 내역을 불러와서 화면에 보여주기 위한 showHistories는 호출되어야 하지 않나요 ??

laco-dev replied:
아 맞습니다 :)

laco-dev commented on app/src/main/java/woowacourse/movie/ui/settings/view/SettingsFragment.kt:
```diff
@@ -1,18 +1,133 @@
         savedInstanceState: Bundle?,
     ): View? {
-        return inflater.inflate(R.layout.fragment_settings, container, false)
+        super.onCreateView(inflater, container, savedInstanceState)
+        _binding = DataBindingUtil.inflate(inflater, R.layout.fragment_settings, container, false)
+        presenter.loadDatabase(bookedTicketDatabase)
+        syncNotificationPermissionWithPrefsAndUI()
+        return binding.root
+    }
```
onCreateView, onViewCreate의 차이는 무엇인가요?
위 로직은 모두 onCreateView에서 해도 괜찮은가요?

yrsel replied:
`onCreateView()` 는 `View`를 연결해주는 생명주기라고 생각되고 위 로직들은 뷰를 연결하는 작업이 아니기에 `onViewCreated()` 가 올바른 위치인 것 같습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/ui/settings/view/SettingsFragment.kt:
```diff
@@ -1,18 +1,133 @@
+         * 설정 변경 시 기존 알림 설정한 데이터 제거 필요
+         * PendingIntent 정보를 room에 저장해야 정확히 비교 후 제거할 수 있을텐데 일단 진행
+         * */
+    }
```
> 질문 3. 알림 제거

데이터가 방대해진다면, 그것은 FCM과 같은 서버 푸시로 바꿔야할 니즈이니, 지금은 고려하지 않아도 괜찮습니다.
데이터베이스 알림 테이블을 관리하는 아이디어 괜찮아 보입니다.

지금 요구사항에서는 영화마다 알림을 설정하는 기능은 없기 때문에 전체 예매 내역을 통해 on/off하는게 자연스럽다고 생각하는데요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/settings/view/SettingsFragment.kt:
```diff
@@ -1,18 +1,133 @@
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            )
+
+        AlarmManagerCompat.setExact(requireContext(), triggerAtMillis, pendingIntent)
```
알림 등록하는 로직이 중복해서 떨궈지지 않았나요? 어떻게 개선해야 할까요?

yrsel replied:
알림 등록하는 로직이 재사용될 확률이 높다는 말씀이실까요 ,,?!
아니면 이 메서드 내에서 중복된 알람이 등록된다는 말씀이실까요 ?!

전자의 경우 Alram에 대한 역할을 수행하는 객체 내부로 이동해서 호출해서 사용하는 방식으로 사용할 수 있을 것 같습니다.
저의 경우 AlarmMangerCompat 이라는 버전 분기를 내부로 숨기는 객체만 존재하고 앱에서 한 번만 스케줄이 등록되어서 따로 수정하지는 않았습니다.

후자의 경우는 어디서 발생할 지 잘 모르겠습니다. 다만 triggerTime이 현재 시간보다 이전일 경우 바로 알람이 울리는 상황에 대한 처리를 진행했습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/settings/view/SettingsPresenter.kt:
```diff
@@ -0,0 +1,33 @@
+
+    override fun loadDatabase(database: BookedTicketDatabase) {
+        db = database
+    }
```
앞서 말씀드린 것 처럼 프레젠터가 View를 인터페이스로 가지고 있는 근본적인 이유에 대해 다시 고민해보셨으면 좋겠어요.
https://woowacourse.slack.com/archives/C08BEPG1SSE/p1746938898246289

yrsel replied:
View, Presenter 모두 데이터베이스에 대해 알지 못하도록 수정하였습니다. 

관련 커밋 
1ae77b5a2ba9099c9c11f6e5fb7706deefc8c862
299577b802a2a77545b29f1602421fe41030813b

laco-dev reviewed (changes requested):
3, 4단계 고생 많으셨습니다.

프레젠터가 왜 View를 인터페이스로 가지고 있는지에 대한 근본적인 학습이 필요해보입니다.
아직은 구현하는 방식이 맞는지 틀린지 모르기 때문에 여러가지 시도를 해 보신 것 같아서 몇가지 잘못된 점도 같이 남겨두었습니다.

yrsel committed:
refactor: MovieBookingActivity movieListFragment 인스턴스 변수 제거

- updateBottomNavigation() : Null 일때 인스턴스 변수 제거 -> 인스턴스 바로 생성

yrsel committed:
refactor: bookingActivity xml 레이아웃 수정

yrsel committed:
refactor: Destination.kt 제거 및 BookingCompleteActivity 뒤로가기 버튼 로직 수정

- MovieBookingActivity 에 대한 프레젠터, 뷰 제거
- Destination 제거

yrsel committed:
refactor: MovieReminderReceiver onReceive 파라미터 타입 수정

yrsel committed:
refactor: MovieBookingActivity 프래그먼트 configuration change 문제 상황 해결

- BookingHistoryFragment가 화면 회전 시 attach 안되는 상황 : BottomNavigation 아이템 선택 리스너에서 현재 프래그먼트가 대체할 프래그먼트와 일치하는 지 체크 (화면을 그릴 때 thread 블럭 내에서 데이터를 가져와야해서 발생한 문제 )

yrsel committed:
refactor: SettingsFragment 데이터 세팅 생명주기 이동

yrsel committed:
rename: BookedTicketConverter 메서드명 수정

yrsel committed:
refactor: BookedTicket id 추가 및 변환 로직 제거

yrsel committed:
refactor: BookedTicketRepository 추상화

yrsel committed:
feat: BookedTicketRepositoryProvider 데이터베이스 생성 담당 객체 추가

yrsel committed:
refactor: BookingActivity 내부 view, presenter, contract 패키지 제거

yrsel committed:
refactor: BookingHistory 내부 view, presenter, contract 패키지 제거

yrsel committed:
refactor: BookingSeatActivity provider 통해서 데이터베이스 얻도록 수정

yrsel committed:
refactor: BookedTicketRepository insert 반환값 제거 및 백그라운드 thread 사용

yrsel committed:
refactor: BookingHistoryFragment provider 통해서 데이터베이스 얻도록 수정

yrsel committed:
feat: SharedProvider sharedPreferences 담당 역할

yrsel committed:
refactor: MovieBookingActivity 메인 화면에서 알림 권한 받도록 수정

yrsel committed:
refactor: BookingSeatActivity 예매 완료 시 조건에 따라 알람 등록

yrsel committed:
refactor: BookingCompleteActivity MVP 구조 수정

yrsel committed:
refactor: BookedTicketRepository insert 시 id값 반환 수정

yrsel committed:
refactor: MovieReminderReceiver 고유한 값 id로 수정

yrsel committed:
refactor: SettingsFragment MVP 구조 제거

yrsel committed:
refactor: BookingHistoryFragment 화면 이동 시, 인자 수정

yrsel committed:
chore: 문자열 리소스 수정

yrsel committed:
feat: StorageProvider 정확한 알람을 제공, 관리하기 위한 상태 추가

yrsel committed:
refactor: MovieReminderReceiver 함수 분리

yrsel committed:
refactor: MovieBookingActivity 정확한 알람 설정

yrsel committed:
refactor: BookingSeatActivity 알람 관련 수정

yrsel committed:
refactor: BookingHistoryFragment 화면 그릴 때 오류 수정

yrsel committed:
refactor: BookingCompletePresenter 코드 위치 수정

yrsel committed:
refactor: AlarmManagerCompat 퍼미션 허용 여부 확인 메서드 추가

yrsel committed:
chore: KtlintFormat 적용

laco-dev reviewed (approved):
고생 많으셨습니다.

이후 미션에서도 지금까지 배운 내용들을 잘 활용해서 많은 것을 배워가셨으면 좋겠습니다.
남은 우테코 기간도 응원하겠습니다 :)
